File: shared/mcp-server-base.js
/**
 * shared/mcp-server-base.js
 *
 * Lightweight MCP server implementation using HTTP + JSON-RPC 2.0.
 * Implements the MCP protocol spec:
 *   - POST /mcp  → JSON-RPC endpoint (initialize, tools/list, tools/call)
 *   - GET  /     → server info
 *
 * Each MCP server extends this base and registers tools.
 */

const express = require('express');
const cors    = require('cors');

class McpServerBase {
  constructor({ name, version = '1.0.0', port }) {
    this.name    = name;
    this.version = version;
    this.port    = port;
    this.tools   = {};         // toolName → { description, inputSchema, handler }
    this.app     = express();
    this.app.use(cors());
    this.app.use(express.json());
    this._setupRoutes();
  }

  // ── Register a tool ───────────────────────────────────────
  registerTool(name, { description, inputSchema, handler }) {
    this.tools[name] = { description, inputSchema, handler };
  }

  // ── MCP JSON-RPC handler ──────────────────────────────────
  _setupRoutes() {
    // Info endpoint
    this.app.get('/', (_, res) => res.json({
      mcp:     true,
      name:    this.name,
      version: this.version,
      tools:   Object.keys(this.tools),
    }));

    // Main MCP JSON-RPC endpoint
    this.app.post('/mcp', async (req, res) => {
      const { jsonrpc, id, method, params } = req.body;

      if (jsonrpc !== '2.0') {
        return res.json({ jsonrpc: '2.0', id, error: { code: -32600, message: 'Invalid Request' } });
      }

      try {
        let result;

        switch (method) {
          // ── initialize ───────────────────────────────────
          case 'initialize':
            result = {
              protocolVersion: '2024-11-05',
              capabilities:    { tools: {} },
              serverInfo:      { name: this.name, version: this.version },
            };
            break;

          // ── tools/list ───────────────────────────────────
          case 'tools/list':
            result = {
              tools: Object.entries(this.tools).map(([name, t]) => ({
                name,
                description:  t.description,
                inputSchema:  t.inputSchema,
              })),
            };
            break;

          // ── tools/call ───────────────────────────────────
          case 'tools/call': {
            const { name: toolName, arguments: toolArgs } = params;
            const tool = this.tools[toolName];

            if (!tool) {
              return res.json({ jsonrpc: '2.0', id, error: { code: -32601, message: `Tool not found: ${toolName}` } });
            }

            const callResult = await tool.handler(toolArgs || {});
            result = {
              content: [{ type: 'text', text: JSON.stringify(callResult, null, 2) }],
              isError: callResult.success === false,
            };
            break;
          }

          // ── notifications/initialized (no response needed) ─
          case 'notifications/initialized':
            return res.json({ jsonrpc: '2.0', id, result: {} });

          default:
            return res.json({ jsonrpc: '2.0', id, error: { code: -32601, message: `Method not found: ${method}` } });
        }

        res.json({ jsonrpc: '2.0', id, result });

      } catch (err) {
        res.json({ jsonrpc: '2.0', id, error: { code: -32603, message: err.message } });
      }
    });
  }

  // ── Start the server ──────────────────────────────────────
  start() {
    this.app.listen(this.port, () => {
      console.log(`  🔧 MCP Server "${this.name}" running on http://localhost:${this.port}`);
    });
  }
}

module.exports = McpServerBase;

File: python_backend/shared/mcp_server_base.py
"""
shared/mcp_server_base.py
FastAPI MCP (Model Context Protocol) server base class.
Implements JSON-RPC 2.0 protocol over HTTP.
"""
import json
import inspect
from typing import Any, Dict, Callable, Awaitable
from fastapi import FastAPI, Request
from pydantic import BaseModel

class JsonRpcRequest(BaseModel):
    jsonrpc: str = "2.0"
    method: str
    params: Dict[str, Any] | None = None
    id: str | int | None = None


def create_mcp_app(server_name: str, version: str, tools_registry: Dict[str, Dict[str, Any]]) -> FastAPI:
    """
    Create a FastAPI MCP server.

    tools_registry format:
    {
        "tool_name": {
            "description": "...",
            "inputSchema": {...},  # JSON Schema
            "handler": async_function_or_sync_function
        }
    }
    """
    app = FastAPI(title=server_name)

    @app.get("/")
    async def server_info():
        """GET / returns server metadata."""
        return {
            "mcp": True,
            "name": server_name,
            "version": version,
            "tools": [
                {
                    "name": tool_name,
                    "description": meta["description"],
                    "inputSchema": meta.get("inputSchema", {}),
                }
                for tool_name, meta in tools_registry.items()
            ],
        }

    @app.post("/mcp")
    async def mcp_handler(req: JsonRpcRequest):
        """POST /mcp handles JSON-RPC 2.0 requests."""
        method = req.method
        params = req.params or {}
        request_id = req.id

        try:
            # Initialize
            if method == "initialize":
                result = {
                    "protocolVersion": "2024-11-05",
                    "capabilities": {"tools": {}},
                    "serverInfo": {"name": server_name, "version": version},
                }

            # List tools
            elif method == "tools/list":
                result = {
                    "tools": [
                        {
                            "name": tool_name,
                            "description": meta["description"],
                            "inputSchema": meta.get("inputSchema", {}),
                        }
                        for tool_name, meta in tools_registry.items()
                    ]
                }

            # Call a tool
            elif method == "tools/call":
                tool_name = params.get("name")
                tool_args = params.get("arguments", {})

                if tool_name not in tools_registry:
                    return {
                        "jsonrpc": "2.0",
                        "error": {
                            "code": -32601,
                            "message": f"Tool '{tool_name}' not found",
                        },
                        "id": request_id,
                    }

                handler = tools_registry[tool_name]["handler"]

                # Call handler (async or sync)
                if inspect.iscoroutinefunction(handler):
                    tool_result = await handler(**tool_args)
                else:
                    tool_result = handler(**tool_args)

                # Wrap result in MCP content format
                result = {
                    "content": [
                        {
                            "type": "text",
                            "text": json.dumps(tool_result, indent=2, default=str),
                        }
                    ],
                    "isError": False,
                }

            # Notifications/initialized
            elif method == "notifications/initialized":
                result = {}

            else:
                return {
                    "jsonrpc": "2.0",
                    "error": {
                        "code": -32601,
                        "message": f"Method '{method}' not found",
                    },
                    "id": request_id,
                }

            # Success response
            return {
                "jsonrpc": "2.0",
                "result": result,
                "id": request_id,
            }

        except Exception as e:
            return {
                "jsonrpc": "2.0",
                "error": {
                    "code": -32603,
                    "message": "Internal error",
                    "data": {"error": str(e)},
                },
                "id": request_id,
            }

    return app
